docs: Add docs for tracking policy activity (#2449)
tigera/docs: calico-enterprise/observability/elastic/policy-activity.mdx

---
description: Learn how to use policy activity logs to gain visibility into policy and rule activity.
---

# Policy activity logs

:::note

This feature is in tech preview. Tech preview features may be subject to significant changes before they become GA.

:::

## About policy activity logs

Policy activity logs provide granular visibility into network policy behavior within your cluster. By tracking exactly which rules are evaluated and triggered, these logs offer policy-centric context that complements standard Flow Logs.

Regularly reviewing policy activity, especially unused policies and rules, helps you maintain a clean and secure cluster by identifying and removing obsolete resources.

> ## Log structure / Data types

Reviewer: ```suggestion
## Log structure
```

| `policy.kind`     | The type of the policy. Possible values: `NetworkPolicy`, `GlobalNetworkPolicy`, `KubernetesNetworkPolicy`, `KubernetesAdminNetworkPolicy`, `KubernetesBaselineAdminNetworkPolicy`, `StagedNetworkPolicy`, `StagedGlobalNetworkPolicy`, `StagedKubernetesNetworkPolicy` | `NetworkPolicy` |
| `policy.namespace`| The namespace where the policy is defined.                                                   | `my-test`                        |
| `policy.name`     | The unique name of the policy.                                                               | `my-tier.my-networkpolicy`       |
| `rule`            | The identifier for the specific rule that was triggered, formatted as `{generation}-{direction}-{rule_index}`. | `2-ingress-4`  |
| `cluster`         | The name of the cluster where the activity occurred.                                         | `cluster`                        |
| `last_evaluated`  | The timestamp when the rule was last evaluated.                                              | `2025-12-01T23:09:28.384714204Z` |

:::note

> The generation number in the `rule` identifier (e.g., the `2` in `2-ingress-4`) reflects the specific version of the policy at the time of evaluation.

Reviewer: I think it's important that the customer knows that this number is a direct reflection of the `metadata.generation` field of the policy. I think "specific version of the policy" is too ambiguous - if the user wants to craft a query for a specific rule identifier, they'll need to know how this version number is resolved

```
POST /tigera_secure_ee_policy_activity*/_search
{
  "size": 0,
  "query": {
    "range": {
      "last_evaluated": {
        "gte": "now-90d"
      }
    }
  },
  "aggs": {
    "active_policies": {
      "composite": {
        "size": 1000,
        "sources": [
          { "kind": { "terms": { "field": "policy.kind" } } },
          { "namespace": { "terms": { "field": "policy.namespace" } } },
          { "name": { "terms": { "field": "policy.name" } } }
        ]
      },
      "aggs": {
        "latest_activity": {
          "max": {
            "field": "last_evaluated"
          }
        }
      }
    }
  }
}
```

To see which rules have been triggered in the last 90 days, use this query:

```
POST /tigera_secure_ee_policy_activity*/_search
{
  "size": 0,
  "query": {
    "range": {
      "last_evaluated": {
        "gte": "now-90d"
      }
    }
  }
}
```

### Step 2: Retrieve all network policies and rules

Export your current policy and rule inventory using `kubectl get ...` commands.

### Step 3: Compare lists

Compare the output from the steps above to identify unused resources:

> - Match: Check if each policy (or specific rule ID) in your Cluster exists in the active lists from Elasticsearch.

Reviewer 2: ```suggestion
- Match: Check if each policy (or specific rule ID) in your cluster exists in the active lists from Elasticsearch.
```

---

> description: Learn how to use Policy Activity Logs to gain visibility into policy and rule activity.

Reviewer 2: Unless we get a strong case from marketing or PM, let's always describe these logs as 'policy activity logs' rather than 'Policy Activity Logs'. A plain description.

---

> # Policy Activity Logs

Reviewer 2: ```suggestion
# Policy activity logs
```

:::note

> This feature is tech preview. Tech preview features may be subject to significant changes before they become GA.

Reviewer: ```suggestion
This feature is in tech preview. Tech preview features may be subject to significant changes before they become GA.
```

Reviewer 2: A quick search shows we're all over the place with this phrasing. @pasanw thanks for pointing this out. I'll come back to this one.

:::

## Big picture

Policy Activity Logs provide granular visibility into network policy behavior within your cluster.

## Value

Track exactly which rules are evaluated and triggered, offering policy-centric context that complements standard Flow Logs.

## How it works

- **Collection**: Calico Enterprise automatically monitors policy and rule evaluations across your cluster.
- **Storage**: These events are securely forwarded and indexed in your Elasticsearch cluster for analysis.

> - **Analysis**: You can query the logs to audit network policy and rules and validate policy enforcement history.

Reviewer 2: I'm inclined to cut this section. There's nothing specific to this type of log. We can presume knowledge of how CE does logs with Elasticsearch.

## Accessing the data

You can access Policy Activity Logs by querying Elasticsearch directly or visualizing them in Kibana (if configured).

### Log structure

Each log entry contains the following key fields:

| Field             | Description                                                                                  | Example                          |
| ----------------- | -------------------------------------------------------------------------------------------- | -------------------------------- |

> | `policy.kind`     | The type of the policy (e.g., `NetworkPolicy`, `GlobalNetworkPolicy`).                       | `NetworkPolicy`                  |

Reviewer: ```suggestion
| `policy.kind`     | The type of the policy. Possible values: `NetworkPolicy`, `GlobalNetworkPolicy`, `KubernetesNetworkPolicy`, `KubernetesAdminNetworkPolicy`, `KubernetesBaselineAdminNetworkPolicy`, `StagedNetworkPolicy`, `StagedGlobalNetworkPolicy`, `StagedKubernetesNetworkPolicy`                  | `NetworkPolicy`                  |
```

| `policy.namespace`| The namespace where the policy is defined.                                                   | `my-test`                        |
| `policy.name`     | The unique name of the policy.                                                               | `my-tier.my-networkpolicy`       |
| `rule`            | The identifier for the specific rule that was triggered, formatted as `{generation}-{direction}-{rule_index}`. | `2-ingress-4`  |
| `cluster`         | The name of the cluster where the activity occurred.                                         | `cluster`                        |
| `last_evaluated`  | The timestamp when the rule was last evaluated.                                              | `2025-12-01T23:09:28.384714204Z` |

:::note

> The generation number in rule identifier represents the specific version of the policy at the time of evaluation.

Reviewer: ```suggestion
The generation number in the `rule` identifier reflects the `metadata.generation` of the policy at the time of evaluation.
```

Reviewer 2: I don't see a reference to `metadata.generation` here or anywhere else in our docs. It would be helpful to explain what this is.

> It increments whenever the policy is updated, allowing you to correlate activity logs with the exact policy configuration that was active when the traffic occurred.

Reviewer: ```suggestion
It increments whenever the spec of the policy is updated, allowing you to correlate activity logs with the exact policy configuration that was active when the traffic occurred.
```

:::

> Supported values for `policy.kind` include:

Reviewer: I think all we need is just the list of possible values - the definitions don't seem necessary for this page. For simplicity I think we can just define the list of values within the existing table. See my suggestion above

Reviewer 2: Agreed. I'm not sure that this table is necessary.

| `KubernetesAdminNetworkPolicy`         | A Kubernetes AdminNetworkPolicy (ANP), allowing cluster administrators to secure the cluster with high-priority rules. |
| `KubernetesBaselineAdminNetworkPolicy` | A Kubernetes BaselineAdminNetworkPolicy (BANP), establishing a default security baseline for the entire cluster.       |
| `StagedNetworkPolicy`                  | A Calico NetworkPolicy in "Staged" mode (logs activity without enforcement).                                           |
| `StagedGlobalNetworkPolicy`            | A Calico GlobalNetworkPolicy in "Staged" mode (logs activity without enforcement).                                     |
| `StagedKubernetesNetworkPolicy`        | A Kubernetes NetworkPolicy that is being previewed in "Staged" mode.                                                   |

### Examples to identify unused policies and rules

Policy Activity Logs record only policies and rules that have been evaluated by actual network traffic. To identify unused policies or rules in your cluster, follow these steps:

> #### Find recently active policies or rules within a period

Reviewer: ```suggestion
#### Find policies active within a recent period
```

   
Use the following Elasticsearch query to list all policies that have been evaluated in the last 90 days:

```
POST /tigera_secure_ee_policy_activity*/_search
{
  "size": 0,
  "query": {
    "range": {
      "last_evaluated": {
        "gte": "now-90d"
      }
    }
  },
  "aggs": {
    "active_policies": {
      "composite": {
        "size": 1000,
        "sources": [
          { "kind": { "terms": { "field": "policy.kind" } } },
          { "namespace": { "terms": { "field": "policy.namespace" } } },
          { "name": { "terms": { "field": "policy.name" } } }
        ]
      },
      "aggs": {
        "latest_activity": {
          "max": {
            "field": "last_evaluated"
          }
        }
      }
    }
  }
}
```

> To see which rules have been triggered in the last 90 days, use this query:

Reviewer: ```suggestion
#### Find rules active within a recent period
To see which rules have been triggered in the last 90 days, use this query:
```

```
POST /tigera_secure_ee_policy_activity*/_search
{
  "size": 0,
  "query": {
    "range": {
      "last_evaluated": {
        "gte": "now-90d"
      }
    }
  }
}
```

#### Compare with your all network policies and rules
- Export your current policy and rule inventory using `kubectl get ...` commands.
- Compare the exported policies/rules with the list of active policies/rules returned by the queries above.

> - Any policies/rules present in your cluster but missing from the activity logs have not been evaluated by traffic in the last 90 days and may be considered unused.

Reviewer: ```suggestion
#### Limitations
- Only rules and policies with evaluations are logged. This means that if a policy or rule is never evaluated, it will not appear in any of the queries above. These queries find rules and policies that _have_ been evaluated, but not within the last 90 days. To find rules and policies that have never been evaluated, you will need to list all rules and policies in your data store, and identify the ones that do not appear in the activity logs.
- Long-lived connections only log a single evaluation when the connection is established, and therefore may show significant periods of time without any logged evaluations. This should be considered when determining if a policy or rule is active based on these logs.
```

I would personally prefer a section that outlines the limitations clearly, that way the user understands the context behind what we suggesting. It doesn't have to be how I've worded it, but this is roughly what I was thinking. Happy to leave it up to you and Chris on what you think is best!

Reviewer 2: I agree that limitations should be called out separately. But is this really a replacement for the "Compare with ..." section?

Author: Yep. I agree with @ctauchen  This section is part of the how to instruction. I can add a separate limitations section.

> > **Tip:** Regularly reviewing unused policies and rules can help you maintain a clean and secure cluster by removing obsolete or unnecessary resources.

Reviewer 2: This seems like the main point of the feature. Perhaps we can remove this tip and include the idea in the intro?

---
description: Learn how to use Policy Activity Logs to gain visibility into policy and rule activity.
---

# Policy Activity Logs

:::note

This feature is tech preview. Tech preview features may be subject to significant changes before they become GA.

:::

## Big picture

Policy Activity Logs provide granular visibility into network policy behavior within your cluster.

## Value

Track exactly which rules are evaluated and triggered, offering policy-centric context that complements standard Flow Logs.

## How it works

> - **Collection**: The Calico Enterprise automatically monitors policy and rule evaluations across your cluster.

Reviewer 3: Nit: s/The Calico/Calico/

| ----------------- | -------------------------------------------------------------------------------------------- | -------------------------------- |
| `policy.kind`     | The type of the policy (e.g., `NetworkPolicy`, `GlobalNetworkPolicy`).                       | `NetworkPolicy`                  |
| `policy.namespace`| The namespace where the policy is defined.                                                   | `my-test`                        |
| `policy.name`     | The unique name of the policy.                                                               | `my-tier.my-networkpolicy`       |
| `rule`            | The identifier for the specific rule that was triggered, formatted as `{generation}-{direction}-{rule_index}`. | `2-ingress-4`  |
| `cluster`         | The name of the cluster where the activity occurred.                                         | `cluster`                        |
| `last_evaluated`  | The timestamp when the rule was last evaluated.                                              | `2025-12-01T23:09:28.384714204Z` |

> ### Sample query in Elasticsearch

Reviewer: Would be nice to also have example queries that show how to evaluate if a given rule/policy is 'unused'

Author: OK. Then hold on the review for my queries. Let me refine them.

> To find policies unused for 90 days:

Reviewer 3: This would find rules that haven't been used in 90d, not whole policies right?

Author: The granularity is rule. So yeah, it's more meaning for usage of a rule, not a policy. For usage on policy level, it requires aggregation, probably some group by query. Let me add more query examples to make it clear.

> To find policies unused for 90 days:

Reviewer: This is slightly misleading:
- It finds rules, not policies
- If a rule/policy was _never_ evaluated, it wouldn't should up in this query even though it is also unused
- If a rule was only ever evaluated once (e.g. more than 90d ago) to establish a long lived connection that is still being used, that rule would be considered unused by this query even though the rule allowed a currently active connection. This scenario is a bit more far-fetched, but could happen for cases like a tunnel connection that is established once and stays up indefinitely

I'm not sure the best way to address this to be honest. We could tweak the wording of this line, or perhaps just add a 'limitations' section afterwards which touches on these details

Author: I added more query examples here. Could you help to check them? Thanks!

---
description: Learn how to use Policy Activity Logs to gain visibility into policy and rule activity.
---

# Policy Activity Logs

:::note

This feature is tech preview. Tech preview features may be subject to significant changes before they become GA.

:::

> ## Big picture

Reviewer: I remember having a discussion with @ctauchen where I believe one of the conclusions is that we are moving away from the "Big picture"/"Value" etc headings. I'll let him chime in on that!

Reviewer 2: You're right, but I've been bad at publicizing that. I'll make suggestions along these lines.

## How it works

- **Collection**: The Calico Enterprise automatically monitors policy and rule evaluations across your cluster.
- **Storage**: These events are securely forwarded and indexed in your Elasticsearch cluster for analysis.
- **Analysis**: You can query the logs to audit network policy and rules and validate policy enforcement history.

## Configuration

Policy Activity Logs are enabled by default.

### Disabling the feature

To disable Policy Activity Logs, set the following environment variables:

- **Felix**: `FELIX_PolicyActivityLogsFileEnabled: false`
- **Fluentd**: `DISABLE_ES_POLICY_ACTIVITY_LOG: true`

## Accessing the data

You can access Policy Activity Logs by querying Elasticsearch directly or visualizing them in Kibana (if configured).

### Log structure

Each log entry contains the following key fields:

| Field             | Description                                                                                  | Example                          |
| ----------------- | -------------------------------------------------------------------------------------------- | -------------------------------- |

> | `policy.kind`     | The type of the policy (e.g., `NetworkPolicy`, `GlobalNetworkPolicy`).                       | `NetworkPolicy`                  |

Reviewer: How will Kubernetes network policy types appear here? I'd recommend showing the full list of potential values given that should be a finite set

| `policy.namespace`| The namespace where the policy is defined.                                                   | `my-test`                        |
| `policy.name`     | The unique name of the policy.                                                               | `my-tier.my-networkpolicy`       |

> | `rule`            | The identifier for the specific rule that was triggered, formatted as `{generation}-{direction}-{rule_index}`. | `2-ingress-4`  |

Reviewer: It's unclear what generation refers to here - it might be best to have a small section that explains the fact that each instance of a policy activity log is tied to a generation-specific instance of a policy rule (and that the generation is from the parent policy)
